src/upload/adapters/nodeRemote.js

```diff
@@ -6,21 +6,77 @@ import url from 'url';
 import decode from '../../utils/decode';
 import local from './nodeLocal';
 
-export default function nodeUrl(client, fileUrl) {
-  return new Promise((resolve, reject) => {
-    tmp.dir((err, dir, cleanupCallback) => {
+export default function nodeUrl(client, fileUrl, options) {
+  let isCancelled = false;
+  let cancel = () => {
+    isCancelled = true;
+  };
+  const promise = new Promise((resolve, reject) => {
+    if (isCancelled) {
+      return reject(new Error('upload aborted'));
+    }
+    return tmp.dir((err, dir, cleanupCallback) => {
       if (err) {
         reject(err);
       }
 
       const encodedFileUrl = decode(fileUrl);
 
+      const cancelTokenSource = axios.CancelToken.source();
+      cancel = () => {
+        isCancelled = true;
+        cancelTokenSource.cancel();
+        cleanupCallback();
+        reject(new Error('upload aborted'));
+      };
+
+      if (isCancelled) {
+        return cancel();
+      }
       return axios({
         url: encodedFileUrl,
         maxRedirects: 10,
-        responseType: 'arraybuffer',
+        responseType: options.onProgress ? 'stream' : 'arraybuffer',
+        cancelToken: cancelTokenSource.token,
+        onDownloadProgress: !options.onProgress
+          ? undefined
+          : event => {
+              const percent = Math.round((event.loaded * 100) / event.total);
+              options.onProgress({
+                type: 'download',
+                payload: { percent },
+              });
+            },
       })
-        .then(response => {
+        .then(async response => {
+          let data;
+          // Axios' onDownloadProgress works only in the browser
+          // so in Node.js we need to implement it ourselves with streams.
+          if (options.onProgress) {
```

Comment:
let me know if you can think of a cleaner way to do this

Reply:
not really an expert on streams here :/